src/us/kbase/workspace/kbase/KBaseWorkspaceConfig.java

```diff
@@ -583,4 +643,198 @@ public String getParamReport() {
+		result = prime * result + ((adminReadOnlyRoles == null) ? 0 : adminReadOnlyRoles.hashCode());
+		result = prime * result + ((adminRoles == null) ? 0 : adminRoles.hashCode());
+		result = prime * result + ((auth2URL == null) ? 0 : auth2URL.hashCode());
+		result = prime * result + ((authURL == null) ? 0 : authURL.hashCode());
+		result = prime * result + ((backendContainer == null) ? 0 : backendContainer.hashCode());
+		// Region doesn't implement hashcode or equals
+		result = prime * result + ((backendRegion == null) ? 0 : backendRegion.id().hashCode());
+		result = prime * result + ((backendToken == null) ? 0 : backendToken.hashCode());
+		result = prime * result + (backendTrustAllCerts ? 1231 : 1237);
+		result = prime * result + ((backendType == null) ? 0 : backendType.hashCode());
+		result = prime * result + ((backendURL == null) ? 0 : backendURL.hashCode());
+		result = prime * result + ((backendUser == null) ? 0 : backendUser.hashCode());
+		result = prime * result + ((bytestreamToken == null) ? 0 : bytestreamToken.hashCode());
+		result = prime * result + ((bytestreamURL == null) ? 0 : bytestreamURL.hashCode());
+		result = prime * result + ((bytestreamUser == null) ? 0 : bytestreamUser.hashCode());
+		result = prime * result + ((db == null) ? 0 : db.hashCode());
+		result = prime * result + ((errors == null) ? 0 : errors.hashCode());
+		result = prime * result + ((handleServiceToken == null) ? 0 : handleServiceToken.hashCode());
+		result = prime * result + ((handleServiceURL == null) ? 0 : handleServiceURL.hashCode());
+		result = prime * result + ((host == null) ? 0 : host.hashCode());
+		result = prime * result + (ignoreHandleService ? 1231 : 1237);
+		result = prime * result + ((infoMessages == null) ? 0 : infoMessages.hashCode());
+		result = prime * result + ((listenerConfigs == null) ? 0 : listenerConfigs.hashCode());
+		result = prime * result + ((mongoPassword == null) ? 0 : mongoPassword.hashCode());
+		result = prime * result + ((mongoUser == null) ? 0 : mongoUser.hashCode());
+		result = prime * result + ((paramReport == null) ? 0 : paramReport.hashCode());
+		result = prime * result + ((sampleServiceToken == null) ? 0 : sampleServiceToken.hashCode());
+		result = prime * result + ((sampleServiceURL == null) ? 0 : sampleServiceURL.hashCode());
+		result = prime * result + ((tempDir == null) ? 0 : tempDir.hashCode());
+		result = prime * result + ((typedb == null) ? 0 : typedb.hashCode());
+		result = prime * result + ((workspaceAdmin == null) ? 0 : workspaceAdmin.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		KBaseWorkspaceConfig other = (KBaseWorkspaceConfig) obj;
+		if (adminReadOnlyRoles == null) {
+			if (other.adminReadOnlyRoles != null)
+				return false;
+		} else if (!adminReadOnlyRoles.equals(other.adminReadOnlyRoles))
+			return false;
```

Comment:
I don't know if there is a way to access these values through `this` -- as it is, you have a lot of repeated stanzas in the form
```
if (variable == null) {
    if (other.variable != null)
        return false;
} else if (!variable.equals(other.variable))
        return false;
}
```
and it would make life easier to be able to iterate through a list of properties, running that check for each of them.

This could also be turned into a static method that compares two objects -- that would make it much easier to do
```
// pseudocode
for (variable in variable_list) {
    if (obj1[variable] == null) {
        if (obj2[variable] != null)
            return false;
    } else if (obj1[variable] != obj2[variable]) {
        return false;
    }
}
```

Unfortunately I am not familiar enough with Java to know if magic of the type I have described here is possible...

Reply:
`equals()` and `hashCode()` are both autogenerated other than the changes I made to get `Region` to work - any changes I made would be blown away the next time I made changes to the state variable and regenerated the functions, so I'd prefer to leave them as is.

Reply:
Sounds good!